pkg/virtual/generic/rootapiserver/root_apiserver.go

```diff
@@ -202,22 +203,39 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 	return s, nil
 }
 
-func getRootHandlerChain(extraConfig *RootAPIExtraConfig, delegateAPIServer genericapiserver.DelegationTarget) func(http.Handler, *genericapiserver.Config) http.Handler {
+func (c completedConfig) getRootHandlerChain(delegateAPIServer genericapiserver.DelegationTarget) func(http.Handler, *genericapiserver.Config) http.Handler {
 	return func(apiHandler http.Handler, genericConfig *genericapiserver.Config) http.Handler {
-		// this is the normal kube handler chain
-		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-			if accepted, prefixToStrip, context := extraConfig.RootPathResolver(req.URL.Path, req.Context()); accepted {
+		return genericapiserver.DefaultBuildHandlerChain(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			if accepted, prefixToStrip, context := c.ExtraConfig.RootPathResolver(req.URL.Path, req.Context()); accepted {
 				req.URL.Path = strings.TrimPrefix(req.URL.Path, prefixToStrip)
 				req.URL.RawPath = strings.TrimPrefix(req.URL.RawPath, prefixToStrip)
 				req = req.WithContext(genericapirequest.WithCluster(context, genericapirequest.Cluster{Name: "virtual"}))
 				delegateAPIServer.UnprotectedHandler().ServeHTTP(w, req)
 				return
 			}
 			http.NotFoundHandler().ServeHTTP(w, req)
-		})
+		}), c.GenericConfig.Config)
 	}
 }
 
+
+var _ genericapirequest.RequestInfoResolver = (*completedConfig)(nil)
+func (c completedConfig) NewRequestInfo(req *http.Request) (*genericapirequest.RequestInfo, error) {
+	defaultResolver := genericapiserver.NewRequestInfoResolver(c.GenericConfig.Config)	
+	if accepted, prefixToStrip, _ := c.ExtraConfig.RootPathResolver(req.URL.Path, req.Context()); accepted {
+		p := strings.TrimPrefix(req.URL.Path, prefixToStrip)
+		rp := strings.TrimPrefix(req.URL.RawPath, prefixToStrip)
+		r2 := new(http.Request)
+		*r2 = *req
```

Comment:
IIRC these are shallow copies, there's a `CloneRequest()` somewhere that you want to use

Reply:
[`Request.Clone`](https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/net/http/request.go;l=372) seems a bit too much for what I need. Here the resulting request is only used as an input for `NewRequestInfo`.
BTW the logic in this code in mainly the same as the one in [`http.StripPrefix`](https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/net/http/server.go;l=2075), apart from the fact that we don't want to manage the handler.